### PR TITLE
Feature/implement complex queries

### PR DESCRIPTION
**Pull request by DominikVoigt**
This PR introduces QueryTransformers for creating structured and fielded queries in the cross-library query language and transforming them into the library-specific query language.

Here a set of Transformers are included, and new transformers can easily be added by extending the `AbstractQueryTransformer`.

Transformers are integrated into the default `performSearch` method of the `SearchBasedFetcher` and all child classes.
Additionally, this PR addresses some bugs in the fetcher tests.

- [ ] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

File: src/main/java/org/jabref/logic/importer/SearchBasedParserFetcher.java [hunk 1]
```diff
@@ -65,9 +62,9 @@ default URL getURLForQuery(ComplexSearchQuery query) throws URISyntaxException,
     /**
      * Constructs a URL based on the query.
      *
-     * @param query the search query
+     * @param transformedQuery the search query
      */
-    URL getURLForQuery(String query) throws URISyntaxException, MalformedURLException, FetcherException;
+    URL getURLForQuery(String transformedQuery, AbstractQueryTransformer transformer) throws URISyntaxException, MalformedURLException, FetcherException;
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/SearchBasedParserFetcher.java:67]: Can you please explain why the `getURL*` methods need to have the transformer as a second argument? It seems that each fetcher has it's own transformer, so why not use a simple class variable?

File: src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java [hunk 2]
```diff
@@ -86,4 +88,9 @@ public void doPostCleanup(BibEntry entry) {
         new FieldFormatterCleanup(StandardField.JOURNAL, new RemoveBracesFormatter()).cleanup(entry);
         new FieldFormatterCleanup(StandardField.TITLE, new RemoveBracesFormatter()).cleanup(entry);
     }
+
+    @Override
+    public AbstractQueryTransformer getQueryTransformer() {
+        return new ZbMathQueryTransformer();
```
Comment (tobiasdiez) [inline src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java:94]: I would put these transformer classes as private inner classes. I think it's an implementation detail how the fetcher transforms the user-query to a query used in the url.

**Review by tobiasdiez — COMMENTED** [review]
Wow, you are now creating huge PRs am laufenden Fließband ;-)

After skimming, I've a few questions about the interface. It feels a bit more complicated than it should ideally be (but than again, that's always the case with these fetcher). Why does the following simplified call chain doesn't work:
- `search(String query)` parsing the user's query and calling `search(parsedQuery)` (This method is mainly there to make testwriting simpler; the parsing should happen also in the GUI to give direct feeback when the query syntax is wrong)
- `search(LuceneQuery query)` calling `getUrl` to obtain the url, then fetch result, parse it, cleanup, return result.
- `getUrl(LuceneQuery query)` generating the url to call from the query. This could use a transformer to convert the query into a string query which is then added to the url via `addParameter("query", transformedQuery);`.  For REST-based API one should prefer to use something like `addParameter("author", query.getAuthor());` instead.

File: src/main/java/org/jabref/logic/importer/SearchBasedParserFetcher.java (hunk 1 above)
Comment (DominikVoigt) [reply src/main/java/org/jabref/logic/importer/SearchBasedParserFetcher.java:67]: The main reason for this is that some transformers require state that is used during query construction (See IEEE) and/or postprocessing (See ArXiv).
I now use a class variable and a reset method. :)

**Comment by DominikVoigt** [comment]
> Wow, you are now creating huge PRs am laufenden Fließband ;-)
> 
> After skimming, I've a few questions about the interface. It feels a bit more complicated than it should ideally be (but than again, that's always the case with these fetcher). Why does the following simplified call chain doesn't work:
> 
> * `search(String query)` parsing the user's query and calling `search(parsedQuery)` (This method is mainly there to make testwriting simpler; the parsing should happen also in the GUI to give direct feeback when the query syntax is wrong)
> * `search(LuceneQuery query)` calling `getUrl` to obtain the url, then fetch result, parse it, cleanup, return result.
> * `getUrl(LuceneQuery query)` generating the url to call from the query. This could use a transformer to convert the query into a string query which is then added to the url via `addParameter("query", transformedQuery);`.  For REST-based API one should prefer to use something like `addParameter("author", query.getAuthor());` instead.

The reason for this is that I wanted to keep the current fetcher method structure.
I don't quite see the advantage of your call chain.
Maybe I don't quite get how your proposed call chain works, could you iterate your concept a bit more in-depth? :)

**Comment by tobiasdiez** [comment]
Take for example the following (new) code:
https://github.com/JabRef/jabref/blob/787f1bcd5e8dbb85cdd85699c22d2c00961757c6/src/main/java/org/jabref/logic/importer/SearchBasedFetcher.java#L18-L46

This has the assumption that each fetcher has an API that supports searching via an extended `query` url parameter, e.g. `query="au: Olly"`. However, there are cases where the API supports passing these extra search fields as extra url parameters, e.g. `author=Olly`. Moreover, I'm not a big fan of using string-based queries when we have more advanced objects (LuceneQuery) that can be used.

So my proposal would be the following rewrite of the above code
```
    /**
     * This method is used to send complex queries using fielded search.
     *
     * @param transformedQuery the search query defining all fielded search parameters
     * @return a list of {@link BibEntry}, which are matched by the query (may be empty)
     */
    List<BibEntry> performSearch(FetcherQuery query) throws FetcherException;
    // In a Parser-Based fetcher this calls getUrlForQuery(query)

    /**
     * Looks for hits which are matched by the given free-text query.
     *
     * @param searchQuery query string that can be parsed into a complex search query
     * @return a list of {@link BibEntry}, which are matched by the query (may be empty)
     */
    default List<BibEntry> performSearch(String query) throws JabRefException {
        return this.performSearch(FetcherQuery.parse(query));
    }

//// Then in a fetcher

   Url getUrlForQuery(FetcherQuery query) {
        urlBuilder.addParamater("query", new xyzTransformer().transform(query))

       // or
      urlBuilder.addParamater("author", query.getAuthor())

   } 
```
For flexibility, I would introduce a `FetcherQuery` class here, deriving from `LuceneQuery` that could provide a few convienentt helper methods.

**Comment by DominikVoigt** [comment]
Okay, I adapted the code accordingly!
The only thing I did not do was to add a FetcherQuery, as it does not seem necessary?
What do you think of my adaption? :)

**Review by tobiasdiez — COMMENTED** [review]
Thanks, I like it a lot!

The idea behind the `FetcherQuery` was to provide access to common query parameters that are context-specific (like start/end year, title, authors).

File: src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java (hunk 2 above)
Comment (tobiasdiez) [reply src/main/java/org/jabref/logic/importer/fetcher/ZbMATH.java]: That shouldn't be a problem. You just move the current code into the fetcher class.

**Review by koppor — APPROVED** [review]

**Pull request merged**